### PR TITLE
refine tree element definitions

### PR DESCRIPTION
**Pull request by pubrrr**

### CODE REVIEW

File: src/domain/expression_tree.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
+#[derive(Debug)]
+pub enum AbstractSyntaxTree {
+    Expression(Expression)
+}
+
+#[derive(Debug)]
+pub enum Expression {
```
Comment (ekinimo) [inline src/domain/expression_tree.rs:7]: we can actually implement parse trait for each sub enum

File: src/domain/expression_tree.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    Sum(Summands),
+    Product(Factors),
+    Atom(Atom),
+}
+
```
Comment (ekinimo) [inline src/domain/expression_tree.rs:12]: why not make Atom the Lvalue ?

File: src/domain/expression_tree.rs (hunk 1 above)
Comment (pubrrr) [reply src/domain/expression_tree.rs:7]: That sounds like a very good idea

File: src/main.rs [hunk 3]
```diff
@@ -23,35 +23,37 @@
...
-use std::io;
-use crate::domain::abstract_syntax_tree::{ExpressionParser, ExpressionLexer};
-use crate::infrastructure::parser::SplitStringParser;
-use crate::infrastructure::lexer::StringLexer;
-
 fn repl() -> () {
     println!("Try some expressions like 1+2, 3*(1+2)");
     loop {
         let mut input = String::new();
         io::stdin().read_line(&mut input).expect("error: unable to read user input");
```
Comment (ekinimo) [inline src/main.rs:44]: maybe there should be meta commands for the repl like :quit that are not valid language constructs

File: src/domain/expression_tree.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+impl Evaluation<i64> for Expression {
+    fn evaluate(&self) -> i64 {
+        match self {
+            Expression::Sum(summands) => add(summands),
+            Expression::Product(factors) => multiply(factors),
+            Expression::Atom(atom) => atom.evaluate(),
+        }
+    }
+}
+
```
Comment (ekinimo) [inline src/domain/expression_tree.rs:41]: cool trick though it wont generalize when Vars enter the game

**Review by ekinimo — APPROVED** [review]

**Pull request merged**